### PR TITLE
[FLINK-28720][hive][connectors] Add Hive partition when flink has no data to write

### PR DESCRIPTION
**Pull request by Tartarus0zm**
## What is the purpose of the change

When inserting a hive static partition, if no data is generated, we also need to add the hive partition

## Brief change log

* Modify the PartitionLoader and FileSystemCommitter to add hive partition when use static partition

## Verifying this change

add ITCase in HiveDialectITCase

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

File: flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/table/FileSystemCommitterTest.java [hunk 1]
```diff
@@ -61,7 +61,12 @@ private void createFile(java.nio.file.Path parent, String path, String... files)
     void testPartition() throws Exception {
```
Comment (luoyuxia) [inline flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/table/FileSystemCommitterTest.java:61]: Also add tests for committing empty partition in here.

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectITCase.java [hunk 2]
```diff
@@ -940,6 +940,100 @@ public void testUnsupportedOperation() {
         }
     }
 
+    @Test
+    public void testInsertIntoNotExistStaticPartitionWithoutData() throws Exception {
```
Comment (luoyuxia) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectITCase.java:944]: I think these tests can move to `HiveTableSinkITCase` and can be composed to one single test.

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/TableEnvHiveConnectorITCase.java [hunk 3]
```diff
@@ -104,12 +104,14 @@ public void testOverwriteWithEmptySource() throws Exception {
                             tableEnv.executeSql("select * from destp order by x").collect());
             assertThat(results.toString()).isEqualTo("[+I[1, 1], +I[2, 2]]");
             // static partitioned table
+            // see FLINK-28720, The semantics of overwrite is to overwrite the original data, so the
```
Comment (luoyuxia) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/TableEnvHiveConnectorITCase.java:107]: I think  `see FLINK-28720,` can be removed.

**Review by luoyuxia — COMMENTED** [review]
@Tartarus0zm Thanks for contribution. I left some comments.

File: flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/PartitionLoader.java [hunk 4]
```diff
@@ -86,6 +86,32 @@ public void loadNonPartition(List<Path> srcDirs) throws Exception {
         overwriteAndMoveFiles(srcDirs, tableLocation);
     }
 
+    /**
+     * The flink job does not write data to the partition, but the corresponding partition needs to
+     * be created.
```
Comment (luoyuxia) [inline flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/PartitionLoader.java:91]: nit:
```suggestion
     * be created or updated.
```

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveTableSinkITCase.java [hunk 5]
```diff
@@ -441,6 +441,61 @@ public void testCustomPartitionCommitPolicy() throws Exception {
         testStreamingWriteWithCustomPartitionCommitPolicy(TestCustomCommitPolicy.class.getName());
     }
 
+    @Test
+    public void testAddEmptyPartition() throws Exception {
```
Comment (luoyuxia) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveTableSinkITCase.java:445]: nit:
```suggestion
    public void testWritingNoDataToPartition() throws Exception {
```

**Review by luoyuxia — COMMENTED** [review]
@Tartarus0zm Thanks for updating. I left some comments.

**Review by luoyuxia — APPROVED** [review]

**Comment by luoyuxia** [comment]
LGTM.

**Comment by luoyuxia** [comment]
@wuchong Could you please help review & merge?

**Review by wuchong — APPROVED** [review]
LGTM
